src/web/cache.rs

```diff
@@ -0,0 +1,170 @@
+use super::STATIC_FILE_CACHE_DURATION;
+use crate::config::Config;
+use iron::{
+    headers::{CacheControl, CacheDirective},
+    AfterMiddleware, IronResult, Request, Response,
+};
+
+#[cfg(test)]
+pub const NO_CACHE: &str = "no-cache, max-age=0";
+
+/// defines the wanted caching behaviour for a web response.
+pub enum CachePolicy {
+    /// no browser or CDN caching.
+    /// In some cases the browser might still use cached content,
+    /// for example when using the "back" button or when it can't
+    /// connect to the server.
+    NoCaching,
+    /// don't cache, plus
+    /// * enforce revalidation
+    /// * never store
+    NoStoreMustRevalidate,
+    /// cache forever in browser & CDN.  
+    /// Valid when you have hashed / versioned filenames and every rebuild would
+    /// change the filename.
+    ForeverInCdnAndBrowser,
+    /// cache forever in CDN, but not in the browser.
+    /// Since we control the CDN we can actively purge content that is cached like
+    /// this, for example after building a crate.
+    /// Example usage: `/latest/` rustdoc pages and their redirects.
+    ForeverOnlyInCdn,
+    /// cache forver in the CDN, but allow stale content in the browser.
+    /// Example: rustdoc pages with the version in their URL.
+    /// A browser will show the stale content while getting the up-to-date
+    /// version from the origin server in the background.
+    /// This helps building a PWA.
+    ForeverInCdnAndStaleInBrowser,
+}
+
+impl CachePolicy {
+    pub fn render(&self, config: &Config) -> Vec<CacheDirective> {
+        match *self {
+            CachePolicy::NoCaching => {
+                vec![CacheDirective::NoCache, CacheDirective::MaxAge(0)]
+            }
+            CachePolicy::NoStoreMustRevalidate => {
+                vec![
+                    CacheDirective::NoCache,
+                    CacheDirective::NoStore,
+                    CacheDirective::MustRevalidate,
+                    CacheDirective::MaxAge(0),
+                ]
+            }
+            CachePolicy::ForeverInCdnAndBrowser => {
+                vec![
+                    CacheDirective::Public,
+                    CacheDirective::MaxAge(STATIC_FILE_CACHE_DURATION as u32),
+                ]
+            }
+            CachePolicy::ForeverOnlyInCdn => {
+                // A missing `max-age` or `s-maxage` in the Cache-Control header will lead to
+                // CloudFront using the default TTL, while the browser not seeing any caching header.
+                // This means we can have the CDN caching the documentation while just
+                // issuing a purge after a build.
+                // https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html#ExpirationDownloadDist
+                vec![CacheDirective::Public]
```

Comment:
typically everything that is cached in the CDN should be `public`, I don't see how this could be an issue? 

In the end we're setting this only when we explicitly want the content to be cached, and we don't set it otherwise.

Reply:
IMO the risk is negligible, especially with the default headers. 

After doing some reading I see your other points and changed it.